cmd/bisync/bisync_test.go

```diff
@@ -0,0 +1,1249 @@
+	files := []string{}
+
+	date, err := time.ParseInLocation(touchDateFormat, dateStr, bisync.TZ)
+	if err != nil {
+		return files, errors.Wrapf(err, "invalid date %q", dateStr)
+	}
+
+	matcher, firstErr := filter.GlobToRegexp(glob, false)
+	if firstErr != nil {
+		return files, errors.Errorf("invalid glob %q", glob)
+	}
+
+	entries, firstErr := f.List(ctx, "")
+	if firstErr != nil {
+		return files, firstErr
+	}
+
+	for _, entry := range entries {
+		obj, isFile := entry.(fs.Object)
+		if !isFile {
+			continue
+		}
+		remote := obj.Remote()
+		if !matcher.MatchString(remote) {
+			continue
+		}
+		files = append(files, dir+remote)
+
+		fs.Debugf(obj, "Set modification time %s", dateStr)
+		err := obj.SetModTime(ctx, date)
+		if err == fs.ErrorCantSetModTimeWithoutDelete {
+			// Workaround for dropbox, similar to --refresh-times
+			err = nil
+			buf := new(bytes.Buffer)
+			size := obj.Size()
+			if size > 0 {
+				err = operations.Cat(ctx, f, buf, 0, size)
+			}
+			info := object.NewStaticObjectInfo(remote, date, size, true, nil, f)
+			if err == nil {
+				_ = obj.Remove(ctx)
+				_, err = f.Put(ctx, buf, info)
+			}
+		}
+		if firstErr == nil {
+			firstErr = err
+		}
+	}
+
+	return files, firstErr
+}
+
+// compareResults validates scenario results against golden dir
+func (b *bisyncTest) compareResults() int {
+	goldenFiles := b.listDir(b.goldenDir)
+	resultFiles := b.listDir(b.workDir)
+
+	// Adapt test file names to their golden counterparts
+	renamed := false
+	for _, fileName := range resultFiles {
+		goldName := b.toGolden(fileName)
+		if goldName != fileName {
+			filePath := filepath.Join(b.workDir, fileName)
+			goldPath := filepath.Join(b.workDir, goldName)
+			require.NoError(b.t, os.Rename(filePath, goldPath))
+			renamed = true
+		}
+	}
+	if renamed {
+		resultFiles = b.listDir(b.workDir)
+	}
+
+	goldenSet := bilib.ToNames(goldenFiles)
+	resultSet := bilib.ToNames(resultFiles)
+	goldenNum := len(goldenFiles)
+	resultNum := len(resultFiles)
+	errorCount := 0
+	const divider = "----------------------------------------------------------"
+
+	if goldenNum != resultNum {
+		log.Print(divider)
+		log.Printf("MISCOMPARE - Number of Golden and Results files do not match:")
+		log.Printf("  Golden count: %d", goldenNum)
+		log.Printf("  Result count: %d", resultNum)
+		log.Printf("  Golden files: %s", strings.Join(goldenFiles, ", "))
+		log.Printf("  Result files: %s", strings.Join(resultFiles, ", "))
+	}
+
+	for _, file := range goldenFiles {
+		if !resultSet.Has(file) {
+			errorCount++
+			log.Printf("  File found in Golden but not in Results:  %s", file)
+		}
+	}
+	for _, file := range resultFiles {
+		if !goldenSet.Has(file) {
+			errorCount++
+			log.Printf("  File found in Results but not in Golden:  %s", file)
+		}
+	}
+
+	for _, file := range goldenFiles {
+		if !resultSet.Has(file) {
+			continue
+		}
+
+		goldenText := b.mangleResult(b.goldenDir, file, false)
```

Comment:
@ncw 

Bisync tests are slightly unusual in relation with rclone test suite.
Instead of creating a test situation programmatically and checking selected result properties
the are based on a simplistic but flexible test scenarios run by a test engine and compare text results.
Therefore, time stamps of test files are predefined. The suite will reject backends without mod-time support.
Regular expressions are used to further make results comparable.
However, I like the principle and thinks it's worth adopting.

Reply:
Probably, mod-time discrepancies with golden results can be ignored under cirtcumstances, like it's done for hashsums, but it needs more testing.

Reply:
`CompareResults` is a very far relative of [fstest.CheckListingWithRoot](https://github.com/rclone/rclone/blob/v1.55.1/fstest/fstest.go#L265) and it bases on `mangleListing` down the code.